Helm v3 - linting multiple charts depends on the order of the charts

<!-- If you need help or think you have found a bug, please help us with your issue by entering the following information (otherwise you can delete this text): -->

Output of `helm version`:
```
version.BuildInfo{Version:"v3.0.0-alpha.2", GitCommit:"97e7461e41455e58d89b4d7d192fed5352001d44", GitTreeState:"clean", GoVersion:"go1.12.7"}
```

Output of `kubectl version`:

N/A

Cloud Provider/Platform (AKS, GKE, Minikube etc.):  

N/A

----
When linting multiple charts using helm3 the order that the charts are passed in matters. The following was done using the, at time of writing, latest commit on the master branch of the charts repo https://github.com/helm/charts/commit/8fb57b472e7a9d5117734566146f272ebc071aa2.

```
$ helm3 lint stable/atlantis stable/apm-server
2 chart(s) linted, 0 chart(s) failed
```

```
$ helm3 lint stable/apm-server stable/atlantis
Error: 2 chart(s) linted, 1 chart(s) failed
	%!s(<nil>)
	[ERROR] templates/: render error at (atlantis/templates/configmap-config.yaml:13:27): wrong type for value; expected string; got map[string]interface {}
```

Using the latest version of helm 2 (v2.14.2) this works as expected
```
$ helm lint stable/apm-server stable/atlantis
==> Linting stable/apm-server
Lint OK

==> Linting stable/atlantis
Lint OK

2 chart(s) linted, no failures
```

```
$ helm lint stable/atlantis stable/apm-server
==> Linting stable/atlantis
Lint OK

==> Linting stable/apm-server
Lint OK

2 chart(s) linted, no failures
```

After reproducing the issue I have been trying to debug this for the past few days.
```
➜  helm3test helm lint ./apm-server ./atlantis
Error: 2 chart(s) linted, 1 chart(s) failed
        %!s(<nil>)
        [ERROR] templates/: template: atlantis/templates/configmap-config.yaml:13:27: executing "atlantis/templates/configmap-config.yaml" at <4>: wrong type for value; expected string; got map[string]interface
{}

➜  helm3test helm lint ./atlantis ./apm-server
2 chart(s) linted, 0 chart(s) failed
```
@karuppiah7890 Seems lke the issue is around https://github.com/helm/helm/blob/dev-v3/pkg/action/lint.go#L61-L73

I checked if there is anything wrong within `lintChart` because of which this might be taking place but had no luck finding anything. After checking with the `2.14` branch I can tice that there are a few changes regarding changing the datatype of vals from `[]byte` to `map[string]inteface{}` which might be important as the error I am seeing wrt 
`[ERROR] templates/: template: atlantis/templates/configmap-config.yaml:13:27: executing "atlantis/templates/configmap-config.yaml" at <4>: wrong type for value; expected string; got map[string]interface` 
might give us some idea. I could not find where `<4>` or rather did not understand where to find it.

After checking out https://github.com/helm/helm/commit/2b81eea1e289da0c0aa1f2f1ca1dca705fa07f70#diff-a3b833d32c8c6c71b4cfd64cf3cdce79 Seems like the answer might lie here. I am a little lost right now and don't know if I am looking in the right direction exactly. Any kind of help would be appreciated.

--- a/pkg/action/action_test.go
+++ b/pkg/action/action_test.go
@@ -166,6 +166,13 @@ func buildChart(opts ...chartOption) *chart.Chart {
 	return c.Chart
 }
 
+func withSampleValues() chartOption {
+	values := map[string]interface{}{"someKey": "someValue"}
+	return func(opts *chartOptions) {
+		opts.Values = values
+	}
+}
+
 func withNotes(notes string) chartOption {
 	return func(opts *chartOptions) {
 		opts.Templates = append(opts.Templates, &chart.File{

--- a/pkg/action/install_test.go
+++ b/pkg/action/install_test.go
@@ -74,6 +74,33 @@ func TestInstallRelease(t *testing.T) {
 	is.Equal(rel.Info.Description, "Install complete")
 }
 
+func TestInstallReleaseWithValues(t *testing.T) {
+	is := assert.New(t)
+	instAction := installAction(t)
+	userVals := map[string]interface{}{}
+	expectedUserValues := map[string]interface{}{}
+	res, err := instAction.Run(buildChart(withSampleValues()), userVals)
+	if err != nil {
+		t.Fatalf("Failed install: %s", err)
+	}
+	is.Equal(res.Name, "test-install-release", "Expected release name.")
+	is.Equal(res.Namespace, "spaced")
+
+	rel, err := instAction.cfg.Releases.Get(res.Name, res.Version)
+	is.NoError(err)
+
+	is.Len(rel.Hooks, 1)
+	is.Equal(rel.Hooks[0].Manifest, manifestWithHook)
+	is.Equal(rel.Hooks[0].Events[0], release.HookPostInstall)
+	is.Equal(rel.Hooks[0].Events[1], release.HookPreDelete, "Expected event 0 is pre-delete")
+
+	is.NotEqual(len(res.Manifest), 0)
+	is.NotEqual(len(rel.Manifest), 0)
+	is.Contains(rel.Manifest, "---\n# Source: hello/templates/hello\nhello: world")
+	is.Equal("Install complete", rel.Info.Description)
+	is.Equal(expectedUserValues, rel.Config)
+}
+
 func TestInstallReleaseClientOnly(t *testing.T) {
 	is := assert.New(t)
 	instAction := installAction(t)

--- a/pkg/action/lint.go
+++ b/pkg/action/lint.go
@@ -83,10 +83,6 @@ func (l *Lint) Run(paths []string, vals map[string]interface{}) *LintResult {
 func lintChart(path string, vals map[string]interface{}, namespace string, strict bool) (support.Linter, error) {
 	var chartPath string
 	linter := support.Linter{}
-	currentVals := make(map[string]interface{}, len(vals))
-	for key, value := range vals {
-		currentVals[key] = value
-	}
 
 	if strings.HasSuffix(path, ".tgz") {
 		tempDir, err := ioutil.TempDir("", "helm-lint")
@@ -120,5 +116,5 @@ func lintChart(path string, vals map[string]interface{}, namespace string, stric
 		return linter, errLintNoChart
 	}
 
-	return lint.All(chartPath, currentVals, namespace, strict), nil
+	return lint.All(chartPath, vals, namespace, strict), nil
 }

--- a/pkg/chartutil/coalesce.go
+++ b/pkg/chartutil/coalesce.go
@@ -34,13 +34,13 @@ import (
 //	- A chart has access to all of the variables for it, as well as all of
 //		the values destined for its dependencies.
 func CoalesceValues(chrt *chart.Chart, vals map[string]interface{}) (Values, error) {
-	if vals == nil {
-		vals = make(map[string]interface{})
+	// create a copy of vals and then pass it to coalesce
+	// and coalesceDeps, as both will mutate the passed values
+	valsCopy := copyMap(vals)
+	if _, err := coalesce(chrt, valsCopy); err != nil {
+		return valsCopy, err
 	}
-	if _, err := coalesce(chrt, vals); err != nil {
-		return vals, err
-	}
-	return coalesceDeps(chrt, vals)
+	return coalesceDeps(chrt, valsCopy)
 }
 
 // coalesce coalesces the dest values and the chart values, giving priority to the dest values.

--- a/pkg/chartutil/coalesce_test.go
+++ b/pkg/chartutil/coalesce_test.go
@@ -19,6 +19,8 @@ package chartutil
 import (
 	"encoding/json"
 	"testing"
+
+	"github.com/stretchr/testify/assert"
 )
 
 // ref: http://www.yaml.org/spec/1.2/spec.html#id2803362
@@ -48,11 +50,20 @@ pequod:
 `)
 
 func TestCoalesceValues(t *testing.T) {
+	is := assert.New(t)
 	c := loadChart(t, "testdata/moby")
 
 	vals, err := ReadValues(testCoalesceValuesYaml)
 	if err != nil {
 		t.Fatal(err)
+	}
+
+	// taking a copy of the values before passing it
+	// to CoalesceValues as argument, so that we can
+	// use it for asserting later
+	valsCopy := make(Values, len(vals))
+	for key, value := range vals {
+		valsCopy[key] = value
 	}
 
 	v, err := CoalesceValues(c, vals)
@@ -102,6 +113,9 @@ func TestCoalesceValues(t *testing.T) {
 			t.Errorf("Expected key %q to be removed, still present", nullKey)
 		}
 	}
+
+	// CoalesceValues should not mutate the passed arguments
+	is.Equal(valsCopy, vals)
 }
 
 func TestCoalesceTables(t *testing.T) {

